### PR TITLE
[lgcdis] Support disassembling table jumps.

### PR DESCRIPTION
**Pull request by kosarev**
The change factors out the symbol-related logic so new symbols can be
generated as we get through section code.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Comment by kosarev** [comment]
Updated to fix formatting as the bots like.

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by kosarev** [comment]
> retest this please

@JaxLinAMD Looks like it didn't help?

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:73]: Could you add some documentation explaining what these `struct`s are for and why we need them?

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:90]: Also here

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:92]: It would be nice to have unit tests for this class and be able to quickly see how it works.

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:97]: Can we remove the negation and make the condition more straightforward? Or even factor out name generation into a separate helper function.

**Comment by Flakebi** [comment]
The `retest this please` or `ok to test` are commands that start Jenkins testing, so there’s nothing you need to do :)

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:298]: nit: could you add inline comments with function argument names to make this more readable?

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:312]: I guess offset and address are synonymous here? Wouldn't it be better to stick to either offset or address in this whole PR to keep things consistent and unambiguous?

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:402]: Why 32? How did you pick the small size?

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:444]: I don't see any insertions into this vector at this point in the function, are there any?

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:388]: This function is very long. Could we split it up into smaller functions before refactoring?

File: lgc/test/lgcdis-table_jump.lgc
Comment (kuhar) [inline lgc/test/lgcdis-table_jump.lgc:1]: I don't see the disassembler file in the coverage report. Are we sure that this test is actually executed?

**Review by kuhar — DISMISSED** [review]

File: lgc/test/lgcdis-table_jump.lgc
Comment (kosarev) [reply lgc/test/lgcdis-table_jump.lgc:1]: The test count seems correct, but this warning looks a bit suspicious. Any chance the coverage data is not being properly updated?

https://github.com/GPUOpen-Drivers/llpc/runs/5756068094?check_suite_focus=true
```
warning: 79 functions have mismatched data
```

**Comment by kosarev** [comment]
Regarding the bot jobs coughing, which I understand is known issue, are there plans/activities towards fixing them?

**Comment by Flakebi** [comment]
> Regarding the bot jobs coughing, which I understand is known issue, are there plans/activities towards fixing them?

Which bots are you referring to?
Jenkins, the coverage reports or the GitHub Actions CI in general (everything but Jenkins)?

**Comment by amdvlk-admin** [comment]
### Test summary for commit b5b524e72fc92f50a44f091697ec73171922189b
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 36c5637649089bf65d42c36b431f47c170b57276</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev eb9d3792bdfda0b6fb94af6bd9989cad51842403</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/201585)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38010/67195 (56.6%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 29185/67195 (43.4%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 37848/67195 (56.3%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 29347/67195 (43.7%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39069/67195 (58.1%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 28126/67195 (41.9%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by kosarev** [comment]
> Which bots are you referring to?

I mean the 'Some checks haven’t completed yet' status hanging forever with the 'Jenkins Pipeline' job seemingly never receiving whatever it expects, like in #1606 at this very moment, causing people to restart the jobs manually time after time.

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [reply lgc/disassembler/Disassembler.cpp:316]: Ack.

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [reply lgc/disassembler/Disassembler.cpp:406]: Thanks for the explanation. Can you leave this as a comment/sidenote in the code?

File: lgc/test/lgcdis-table_jump.lgc
Comment (vettoreldaniele) [reply lgc/test/lgcdis-table_jump.lgc:1]: The coverage report does not currently cover LLVM due to size limitations. Since the disassembler is used as a LLVM library, it will not appear in the report.

Comment (kuhar) [reply lgc/test/lgcdis-table_jump.lgc:1]: @vettoreldaniele :^

Comment (kuhar) [reply lgc/test/lgcdis-table_jump.lgc:1]: @vettoreldaniele would it be possible to include lgc but not the rest of llvm? How difficult would that be?

Comment (vettoreldaniele) [reply lgc/test/lgcdis-table_jump.lgc:1]: After checking, yes it is possible. This is outside of LLVM so it will work. The reason it's not included at the moment is that we simply don't have `lgcdis` in the binaries analyzed by the coverage report generator. We only have `amdllpc` currently. I will submit a patch to add this. We should probably include other binaries that are run in these tests to make sure they're accounted for. Anything outside LLVM is easy to handle.

Comment (kosarev) [reply lgc/test/lgcdis-table_jump.lgc:1]: Looks like it's #1756.

File: lgc/disassembler/Disassembler.cpp
Comment (kosarev) [reply lgc/disassembler/Disassembler.cpp:95]: OK, I will degenerate the class into a function for now then.

**Comment by kosarev** [comment]
Updated to not have the symbol pool as a class. Please take another look.

File: lgc/disassembler/Disassembler.cpp
Comment (kosarev) [reply lgc/disassembler/Disassembler.cpp]: Done.

**Comment by amdvlk-admin** [comment]
### Test summary for commit ac81ecd3215cae7e91948f0c69c7040952688052
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 1ff58cafbae03aa571b0dde83aefd0d1fdfb8a9b</li>
<li>SPVGEN: dev 1b6b362984a2d99d897a7edd2e90cb230c933b87</li>
<li>XGL: dev 685edbe4a2a6dc2051820498d414ca650261dde8</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 007e2c31e995541d6c0a91db17044e3da946c6ed</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/188328)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38110/67195 (56.7%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 29085/67195 (43.3%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33664/53938 (62.4%)</li>
<li><details><summary>Failed: 1/53938 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.arithmetic.compute.subgroupexclusivemin_u64vec4
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 20273/53938 (37.6%)</li><li>Warnings: 0/53938 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39169/67195 (58.3%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 28026/67195 (41.7%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by kosarev** [comment]
retest this please

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:792]: Can we have some unit tests?

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:609]: nit: llvm prefers early exits

File: lgc/disassembler/Disassembler.cpp
Comment (kuhar) [inline lgc/disassembler/Disassembler.cpp:571]: Could we have some unit tests for this?

**Review by kuhar — COMMENTED** [review]
The code looks fine to me but deserves more thorough testing IMO.
For example, most mnemonics in `disasmTableJump` seem not exercised: https://storage.googleapis.com/amdvlk-llpc-github-ci-artifacts-public/coverage_release_clang_shadercache_coverage_assertions_2088609165/coverage/vulkandriver/drivers/llpc/lgc/disassembler/Disassembler.cpp.html#L512

**Comment by amdvlk-admin** [comment]
### Test summary for commit 6c852a27cf60c3850da6a1608916aaaa12bcbb0b
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 1ff58cafbae03aa571b0dde83aefd0d1fdfb8a9b</li>
<li>SPVGEN: dev 1b6b362984a2d99d897a7edd2e90cb230c933b87</li>
<li>XGL: dev 685edbe4a2a6dc2051820498d414ca650261dde8</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 007e2c31e995541d6c0a91db17044e3da946c6ed</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/201585)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38110/67195 (56.7%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 29085/67195 (43.3%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 37948/67195 (56.5%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 29247/67195 (43.5%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39169/67195 (58.3%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 28026/67195 (41.7%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by kuhar** [comment]
> > The code looks fine to me but deserves more thorough testing IMO. For example, most mnemonics in `disasmTableJump` seem not exercised: https://storage.googleapis.com/amdvlk-llpc-github-ci-artifacts-public/coverage_release_clang_shadercache_coverage_assertions_2088609165/coverage/vulkandriver/drivers/llpc/lgc/disassembler/Disassembler.cpp.html#L512
> 
> The non-covered `return false`s mean we don't have tests for all sorts of sequences that are not table jumps. Adding such tests would mean having a number of similar lit tests of which usefulness is not immediately obvious. So once again, please can you expand on how many tests you see enough here and how you see them useful. Thanks.

From the perspective of a person who is not familiar with this part of the code base, any tests that ensure I or others don't accidentally break this code would be highly appreciated. This is especially important for refactoring or merging with the internal code base that tends to introduce spurious changes. Code that's not exercised tends to bitrot. For example, there's a condition that looks non-trivial to me right now and I wouldn't know how to write an input that would allow me to check it:
```
  if (addc.mnemonic != "s_addc_u32" || *addc.op0.sReg != *add.op0.sReg + 1 || *addc.op1.sReg != *load.op0.sReg + 1 ||
      *addc.op2.sReg != *getpc.op0.sReg + 1)
```

I'm not saying that lack of unit tests is a blocker for this change, but I see it as an alternative to lit tests that can make it easier to keep the code exercised and working without having to write large and repetitive lit tests.

**Comment by kosarev** [comment]
> From the perspective of a person who is not familiar with this part of the code base, any tests that ensure I or others don't accidentally break this code would be highly appreciated. This is especially important for refactoring or merging with the internal code base that tends to introduce spurious changes. Code that's not exercised tends to bitrot. For example, there's a condition that looks non-trivial to me right now and I wouldn't know how to write an input that would allow me to check it:
> 
> ```
>   if (addc.mnemonic != "s_addc_u32" || *addc.op0.sReg != *add.op0.sReg + 1 || *addc.op1.sReg != *load.op0.sReg + 1 ||
>       *addc.op2.sReg != *getpc.op0.sReg + 1)
> ```

If we agree that LLVM may serve as an example of such a code which functionality is sentenced to survive a great deal of changes of all sorts on daily basis, then this experience could probably work as evidence for regression tests not having the comprehensiveness of conformance testing being sufficient (and even desirable, for their speed) for our usual dev-tool needs.

In this particular case we have a test to make sure table jumps are still recognised as such and all other instruction sequences work as negative tests for the logic. Then if we have a bug slipped through, then there's going to be another (one) regression test inspired by that new knowledge that we can gain from investigating it regarding what are those ways in which we may likely have problems with this code, thus probably suggesting ideas about how to prevent same kind of problems in the future. And then regression tests will do what they are probably supposed to -- making sure we don't run into the same issue again. As to full-sizely ensuring the correctness of a code, I'm afraid that is a bit beyond what we as humanity are practically capable of, even in these blessed times when unit tests are already invented.

File: lgc/test/CMakeLists.txt
Comment (Flakebi) [inline lgc/test/CMakeLists.txt:45]: I think `llvm-mc` can be added to the `LGC_TEST_DEPENDS` list, then it doesn’t need to be repeated

Comment (kosarev) [reply lgc/test/CMakeLists.txt]: Regarding `add_lit_testsuites` dependencies, that's `LLVM_TEST_DEPENDS`, not `LGC_TEST_DEPENDS`. I wonder why would we need dependencies for `add_lit_testsuite` and `add_lit_testsuites` be different. Plus whether we actually need the latter now that we don't really have multiple test suites there.

Comment (Flakebi) [reply lgc/test/CMakeLists.txt]: Ah, I missed that. No idea what the `add_lit_testsuites` is for.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 68aeeec91f15a909976d3c7a6c47fbe2ab22cc46
<details><summary>CTS tests (Failed: 1/155928)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38006/67091 (56.6%)</li>
<li>Failed: 0/67091 (0.0%)</li>
<li>Not Supported: 29085/67091 (43.4%)</li><li>Warnings: 0/67091 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 14728/21746 (67.7%)</li>
<li><details><summary>Failed: 1/21746 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.spirv_assembly.type.vec3.i32.bit_field_u_extract_offset32_count64_tessc
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 7017/21746 (32.3%)</li><li>Warnings: 0/21746 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39065/67091 (58.2%)</li>
<li>Failed: 0/67091 (0.0%)</li>
<li>Not Supported: 28026/67091 (41.8%)</li><li>Warnings: 0/67091 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by kosarev** [comment]
retest this please

**Review by kuhar — COMMENTED** [review]
I don't want to block this PR and resign from the review.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 68aeeec91f15a909976d3c7a6c47fbe2ab22cc46
<details><summary>ShaderDB test (Failed: 0/1134)</summary><ul>

<li>Passed: 1130/1134 (99.6%)</li>
<li>Failed: 0/1134 (0.0%)</li>
<li>Not Supported: 4/1134 (0.4%)</li><li>Warnings: 0/1134 (0.0%)</li>
<li>Expected Failed: 0/1134 (0.0%)</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/201273)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38006/67091 (56.6%)</li>
<li>Failed: 0/67091 (0.0%)</li>
<li>Not Supported: 29085/67091 (43.4%)</li><li>Warnings: 0/67091 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 37844/67091 (56.4%)</li>
<li>Failed: 0/67091 (0.0%)</li>
<li>Not Supported: 29247/67091 (43.6%)</li><li>Warnings: 0/67091 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39065/67091 (58.2%)</li>
<li>Failed: 0/67091 (0.0%)</li>
<li>Not Supported: 28026/67091 (41.8%)</li><li>Warnings: 0/67091 (0.0%)</li>
</ul></details>


</ul></details>

**Review by trenouf — APPROVED** [review]
Sorry Ivan, I forgot about this again.

LGTM.

I think there are some valid points about testing here, but equally we don't have the resources to make it completely comprehensive.

**Comment by amdrexu** [comment]
I see @vettoreldaniele and @Flakebi request the change. Need their re-approval to further handle the PR.

**Comment by vettoreldaniele** [comment]
I just commented on his PR on some code coverage-related questions, no need for changes from me.

File: lgc/disassembler/Disassembler.cpp
Comment (kosarev) [reply lgc/disassembler/Disassembler.cpp:366]: Done in https://github.com/GPUOpen-Drivers/llpc/pull/1889.

**Pull request merged**
